src/change_tracker.rs

```diff
@@ -0,0 +1,209 @@
+            world,
+            added: false,
+            changed: false,
+            removed: false,
+        }
+    }
+}
+
+impl<T: Component> Default for ChangeTracker<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+struct Previous<T>(T);
+
+/// Collection of iterators over changes in `T` components
+pub struct Changes<'a, T>
+where
+    T: Component + Clone + PartialEq,
+{
+    tracker: &'a mut ChangeTracker<T>,
+    world: &'a mut World,
+    added: bool,
+    changed: bool,
+    removed: bool,
+}
+
+impl<'a, T> Changes<'a, T>
+where
+    T: Component + Clone + PartialEq,
+{
+    /// Iterate over entities which were given a new `T` component after the preceding
+    /// [`track`](ChangeTracker::track) call, including newly spawned entities
+    pub fn added(&mut self) -> impl ExactSizeIterator<Item = (Entity, &T)> + '_ {
+        self.tracker.added_components.clear();
+        self.added = true;
+        DrainOnDrop(
+            self.tracker
+                .added
+                .query_mut(self.world)
+                .inspect(|&(e, x)| self.tracker.added_components.push((e, x.clone()))),
+        )
+    }
+
+    /// Iterate over `(entity, old, new)` for entities whose `T` component has changed according to
+    /// [`PartialEq`] after the preceding [`track`](ChangeTracker::track) call
+    pub fn changed(&mut self) -> impl Iterator<Item = (Entity, T, &T)> + '_ {
+        DrainOnDrop(
+            self.tracker
+                .changed
+                .query_mut(self.world)
+                .filter_map(|(e, (new, old))| {
+                    (*new != old.0).then(|| {
+                        let old = mem::replace(&mut old.0, new.clone());
+                        (e, old, new)
+                    })
+                }),
+        )
+    }
+
+    /// Iterate over entities which lost their `T` component after the preceding
+    /// [`track`](ChangeTracker::track) call, excluding any entities which were despawned
+    pub fn removed(&mut self) -> impl ExactSizeIterator<Item = (Entity, T)> + '_ {
+        self.tracker.removed_components.clear();
+        self.removed = true;
+        // TODO: We could make this much more efficient by introducing a mechanism for queries to
+        // take ownership of components directly.
+        self.tracker
+            .removed_components
+            .extend(self.tracker.removed.query_mut(self.world).map(|(e, ())| e));
+        DrainOnDrop(
+            self.tracker
+                .removed_components
+                .drain(..)
+                .map(|e| (e, self.world.remove_one::<Previous<T>>(e).unwrap().0)),
+        )
+    }
+}
+
+impl<'a, T: Component> Drop for Changes<'a, T>
+where
+    T: Component + Clone + PartialEq,
+{
+    fn drop(&mut self) {
+        if !self.added {
+            _ = self.added();
+        }
+        for (entity, component) in self.tracker.added_components.drain(..) {
+            self.world.insert_one(entity, Previous(component)).unwrap();
```

Comment:
Not strictly part of this change but something that might be motivated by it: If not queries that add/remove components, maybe a bulk insertion/removal method might be helpful here. I was thinking about sorting the entities by archetype and then handling all entities in a given archetype as a single operation? (This would probably require quite a few extra code paths, so I am not sure about the cost/benefit ratio.)

Reply:
> maybe a bulk insertion/removal method might be helpful here

Yeah, I want this. Repeatedly inserting/removing from the world scope is definitely suboptimal. See also the TODO, and https://github.com/Ralith/hecs/pull/337. A narrower solution than https://github.com/Ralith/hecs/pull/337 might be sufficient to get a nice fast-path here, but I'm not sure how to come up with a good API design for bulk insert where a new value is dependent on existing components. Maybe require a constructor closure, and let the query overwrite it? Bulk removal is easier API-design-wise but probably a nontrivial chunk of code all the same.